src/stream/gst/pipeline_builder.rs

```diff
@@ -0,0 +1,225 @@
+use crate::{
+    stream::types::VideoCaptureConfiguration,
+    video::{
+        types::{VideoEncodeType, VideoSourceType},
+        video_source_gst::VideoSourceGstType,
+    },
+};
+use crate::{
+    stream::webrtc::{
+        signalling_server::DEFAULT_SIGNALLING_ENDPOINT,
+        turn_server::{DEFAULT_STUN_ENDPOINT, DEFAULT_TURN_ENDPOINT},
+        utils::webrtc_usage_hint,
+    },
+    video_stream::types::VideoAndStreamInformation,
+};
+use log::*;
+use simple_error::SimpleError;
+
+#[derive(Clone, Debug, Default)]
+pub struct Pipeline {
+    pub description: String,
+}
+
+impl Pipeline {
+    pub fn new(
+        video_and_stream_information: &VideoAndStreamInformation,
+        configuration: &VideoCaptureConfiguration,
+        is_webrtcsink: bool,
+    ) -> Result<Self, SimpleError> {
+        let encode = configuration.encode.clone();
+        let endpoints = &video_and_stream_information.stream_information.endpoints;
+        let video_source = &video_and_stream_information.video_source;
+
+        let capability = Pipeline::build_capability_string(video_source, &encode, &configuration)?;
+
+        let source = Pipeline::build_pipeline_source(video_source, &capability)?;
+        let transcode = Pipeline::build_pipeline_transcode(video_source, &encode, is_webrtcsink)?;
+        let payload = Pipeline::build_pipeline_payload(&encode, is_webrtcsink)?;
+        let sink = Pipeline::build_pipeline_sink(&endpoints, is_webrtcsink)?;
+
+        let description = format!("{source}{transcode}{payload}{sink}");
```

Comment:
Actually, I think this would be a good point to switch to using the GST API to "make" the elements, and then "link" them, [as they recommend doing](https://gstreamer.freedesktop.org/documentation/application-development/basics/elements.html?gi-language=c#linking-elements) instead of creating the entire pipeline as a string.